File: problems/101-symmetric-tree.ts
function isSymmetric(root: TreeNode | null): boolean {
  let left = [root?.left ?? null]
  let right = [root?.right ?? null]
  let hasNextRow = !!(left[0] || right[0])

  while (hasNextRow) {
    const nextLeft = []
    const nextRight = []
    hasNextRow = false

    for (let i = 0; i < left.length; i++) {
      if (left[i]?.val !== right[i]?.val) {
        return false
      }

      nextLeft.push(...[left[i]?.left ?? null, left[i]?.right ?? null])
      nextRight.push(...[right[i]?.right ?? null, right[i]?.left ?? null])

      if (!hasNextRow) {
        hasNextRow = !!(
          left[i]?.left ||
          left[i]?.right ||
          right[i]?.right ||
          right[i]?.left
        )
      }
    }

    left = nextLeft
    right = nextRight
  }

  return true
}

File: problems/101_symmetric_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: TreeNode | None) -> bool:
        if not root:
            return True
        nodes = [root.left, root.right]
        while nodes:
            n1 = nodes.pop(0)
            n2 = nodes.pop(0)
            if not n1 and not n2:
                continue
            elif not n1 or not n2 or n1.val != n2.val:
                return False
            else:
                nodes.append(n1.left)
                nodes.append(n2.right)
                nodes.append(n1.right)
                nodes.append(n2.left)

        return True
